node/forest_libp2p/src/blocksync/message.rs

```diff
@@ -65,46 +68,88 @@ pub struct TipsetBundle {
-                msgs.push(
-                    values
-                        .get(*idx as usize)
-                        .cloned()
-                        .ok_or_else(|| "Invalid message index".to_string())?,
-                );
-            }
-            Ok(msgs)
-        }
+impl TryFrom<&TipsetBundle> for FullTipset {
+    type Error = String;
 
-        let mut blocks: Vec<Block> = Vec::with_capacity(tsb.blocks.len());
+    fn try_from(tsb: &TipsetBundle) -> Result<FullTipset, Self::Error> {
+        fts_from_bundle_parts(
+            tsb.blocks.clone(),
+            &tsb.bls_msgs,
+            &tsb.secp_msgs,
+            &tsb.bls_msg_includes,
+            &tsb.secp_msg_includes,
+        )
+    }
+}
 
-        for (i, header) in tsb.blocks.into_iter().enumerate() {
-            let bls_messages = values_from_indexes(&tsb.bls_msg_includes[i], &tsb.bls_msgs)?;
-            let secp_messages = values_from_indexes(&tsb.secp_msg_includes[i], &tsb.secp_msgs)?;
+fn fts_from_bundle_parts(
+    headers: Vec<BlockHeader>,
+    bls_msgs: &[UnsignedMessage],
+    secp_msgs: &[SignedMessage],
+    bls_msg_includes: &[Vec<u64>],
+    secp_msg_includes: &[Vec<u64>],
+) -> Result<FullTipset, String> {
+    // TODO: we may already want to check this on construction of the bundle
+    if headers.len() != bls_msg_includes.len() || headers.len() != secp_msg_includes.len() {
+        return Err(
+            "Invalid formed Tipset bundle, lengths of includes does not match blocks".to_string(),
+        );
+    }
 
-            blocks.push(Block {
-                header,
-                secp_messages,
-                bls_messages,
-            });
+    fn values_from_indexes<T: Clone>(indexes: &[u64], values: &[T]) -> Result<Vec<T>, String> {
+        let mut msgs = Vec::with_capacity(indexes.len());
+        for idx in indexes.iter() {
+            msgs.push(
+                values
+                    .get(*idx as usize)
+                    .cloned()
+                    .ok_or_else(|| "Invalid message index".to_string())?,
+            );
         }
+        Ok(msgs)
+    }
+
+    let mut blocks: Vec<Block> = Vec::with_capacity(headers.len());
 
-        Ok(FullTipset::new(blocks).map_err(|e| e.to_string())?)
+    for (i, header) in headers.into_iter().enumerate() {
+        let message_count = bls_msg_includes[i].len() + secp_msg_includes[i].len();
+        if message_count > BLOCK_MESSAGE_LIMIT {
+            return Err(format!(
+                "Block {} in bundle has too many messages ({} > {})",
+                i, message_count, BLOCK_MESSAGE_LIMIT
+            ));
+        }
+        let bls_messages = values_from_indexes(&bls_msg_includes[i], &bls_msgs)?;
+        let secp_messages = values_from_indexes(&secp_msg_includes[i], &secp_msgs)?;
+
+        blocks.push(Block {
+            header,
+            secp_messages,
+            bls_messages,
+        });
     }
+
+    Ok(FullTipset::new(blocks).map_err(|e| e.to_string())?)
```

Comment:
Optional : another approach would be to move this into a try_fold but the downside is that you would be allocating

Reply:
for what benefit?

Reply:
You would be potentially getting rid of a mutable block vec and switching control to the try_fold but the scope if fairly small so not too worried about that

Reply:
why would you try fold instead of just doing an iter() map()? I'm not really seeing any optimization here, but I can commit this?

Reply:
That's fair, I can see that. 

Will retract my suggestion